README.md
# HoneyGainGrapher
## Installation
  1) Install npm and nodejs https://nodejs.org/en/
  2) Install python 3.x https://www.python.org/downloads/
  3) Download this repository <br> with Git `git clone https://github.com/chaos511/HoneyGainGrapher`<br>On windows download zip and extract
  4) Install all the node dependencies with `npm install` <br> On windows double click install
  5) Add auth token to config file see Finding Your Auth Token for details
  6) Start The program with `node index.js` or double click start
  7) Go to http://127.0.0.1 with any browser
## Finding Your Auth Token
  1) Go to http://https://dashboard.honeygain.com/ and login
  2) Open Devtools (in chrome you can do this with ctrl+shift+i)
  3) Go to the network tab in Devtools
  4) Click on the request for balances and look for a section labeled "Request Headers" and under that, a section labeled "authorization"
  ![Alt text](/Capture2.PNG?raw=true )
  5) Copy the auth token from the authorization section starting after the word Bearer and past it into the file config/default.json
  
## Features
  Adjustable ping interval<br>
  Supports custom device titles (on honeygains website)<br>
  Supports Multipage Active Devices lists(more than 10 devices)<br>
  Dashboard with Current balance, previous earnings, and next estimated payout<br>
  Pool mode where multiple devices can summed into one user<br>
  Supports https
  TODO: Add support for password login<br>
  TODO: Drops sample points when data is older than x days to save space<br>
  TODO: Add compression to http server<br>
  TODO: Runs in the background as a service<br>
## Config File
  The config file is located at config/default.json the felds in the config file are
   
  authToken: This is where you need to put your honeygain auth token

  pingInterval: How offen to record data
  
  startOnTheHour: If the ping interval should start counting from on the hour example x:00:00 instead of when the program starts

  graphTitle: The title to be showen above the graph on the  webpage

  webserverPort: The port the webserver should sisten on

  webserverHost: The ip address the webserver should sisten on


## Image
![Alt text](/Capture.PNG?raw=true )
![Alt text](/Capture3.PNG?raw=true )

  
  
  


ChangeLog.txt
v1.1.0 Add Support For More than 10 Devices
v1.1.1 Fix uri encoding for device names
v1.1.2 Fix graph showing UTC instead of local time
v1.1.3 Show custom title in graph instead of manufacturer
v1.1.4 Fix using js if statement in python
v1.1.5 Fix null graph names and graph not showing all device
v1.2.0 Allow setting graph name in config file
v1.3.0 Make ping interval start on the hour instead of when the program starts
v1.3.1 Fix only pulling last page after first interval
v1.4.0 Sort traces in graph by data usage at last update interval
v1.5.0 Add Dashboard 
v1.6.0 Add graph tab to dashboard
v1.6.1 Make dashboard default to overview tab
v1.6.2 Fix dashboard links redirecting to absolute url
v1.6.3 Move seven day logic to sever-side to save bandwidth
v1.6.4 Fx total trace sometimes not showing
v1.7.0 Add devices tab to dashboard and change how data is stored ~%50 space saved
v1.7.1 Fix endless loop of getdevicebalance
v1.7.2 Fix getdevicebalance being called for every device
v1.7.3 DecodeURI the title in the device table and add export csv button
v1.8.0 Add users tab to dashboard
v1.8.1 Fix not setting min date in date selector
v1.8.2 Improved error handling for unparseable json messages
v1.9.0 Add Last earning column and device filter to device table and separate html/js files
v1.10.0 Add Last earning column to user table 
v1.11.0 Add option to disable graph
v1.11.1 Code Cleanup and Enable DeepScan
v1.12.0 Add end date selector for both user and device tables
v1.12.1 Code Cleanup
v1.12.2 Fix last earning is undefined when end date is changed
v1.12.3 Fix tables sometimes not displaying correctly
v1.13.0 Add Total Balance and Balance Gained column to user and device tables
v1.13.1 Fix showing Credits Gained twice
v1.13.2 Fix Total Credits having $ prefix
v1.13.3 Fix showing loading... when Balance or last7day earnings is zero
v1.14.0 Add Transactions Page
v1.14.1 Fix Display mode option not working on devices tab
v1.14.2 Add referral to transaction type list
v1.14.3 Add payout_reservation_return and payout type list
v1.14.4 Rename selector Only Payout from Only Payout Reservation
v1.14.5 Include payout_reservation_return in total payout calculation
v1.14.6 Fix transaction showing multiple times and add more info to ws error screen
v1.15.0 make dashboard the default web page
v1.15.1 Change WS type to null
v1.15.2 Accept all ws connections
v1.16.0 Change all usage of webSockets to http get requests
v1.16.1 Add v param to scripts for cache control
v1.17.0 Add last hour earnings to overview page
v1.17.1 Add v param to styleSheet for cache control
v1.17.2 Enable get balance
v1.17.3 Fix graph showing incorrect values when sample points are not one hour apart
v1.18.0 Add support for https
v1.18.1 Fix skipping page two of devices
v1.18.2 Change font to Monda
v1.18.3 Change overview layout
v1.19.0 Add more stuff to overview page of the dashboard
v1.19.1 Fix Last Hour showing negative at UTC 0:00
v1.19.2 encode ' in idmap.json as %27
v1.19.3 Start adding login system

genidmap.js
const fs = require('fs');

function wsReadFile(fpath){
    return new Promise(resolve => {
        fs.readFile(fpath, function(error, content) {
            if (error) {
                console.log("Error 404: File "+fpath+" Not Found");
                resolve("{}")
            }
            else {
                resolve(content)
            }
        });
    })
}
async function main(){
    var content=await wsReadFile("data.json")
    jsonData=JSON.parse('['+content.slice(0,-2)+']')
    var dataBuffer=""
    var idmap
    try{
        idmap=JSON.parse(await wsReadFile("idmap.json","{}"))
    }catch{

    }
    if(idmap==undefined){
        idmap={}
    }
    for(device in jsonData){
        if(idmap[jsonData[device]['1']]==undefined||idmap[jsonData[device]['1']].title!=(jsonData[device]['4'])){
            idmapData={
                id:jsonData[device]['1'],
                manufacturer:(jsonData[device]['2']),
                model:(jsonData[device]['3']),
                title:(jsonData[device]['4']),
                platform:(jsonData[device]['5']),
                version:(jsonData[device]['6']),
            }
                console.log("device not found or changed adding: "+JSON.stringify(idmapData))
            
            idmap[jsonData[device]['1']]=idmapData
        }
        deviceData={
            1:jsonData[device]['1'],
            // 2:encodeURI(jsonData[device].['2']),
            // 3:encodeURI(jsonData[device].['3']),
            // 4:encodeURI(jsonData[device].['4']),
            // 5:encodeURI(jsonData[device].['5']),
            // 6:encodeURI(jsonData[device].['6']),
            7:jsonData[device]['7'],
            8:jsonData[device]['8'],
            9:jsonData[device]['9'],
        };

        dataBuffer+=JSON.stringify(deviceData)+",\n"
    }

    fs.writeFile("data.json", dataBuffer, function(err) {
        if(err) {
            return console.log(err);
        }
    }); 
    fs.writeFile("idmap.json", JSON.stringify(idmap), function(err) {
        if(err) {
            return console.log(err);
        }
    }); 
}
main()

www/helper.js
function unixToYYYYMMDD(unixtime){
    var date = new Date(unixtime);
    var dd = date.getDate();
    var mm=date.getMonth()+1
    if(dd<10){dd='0'+dd} 
    if(mm<10){mm='0'+mm} 
    return date.getFullYear()+'-'+mm+'-'+dd;
}
function getHoursFromMidnight(){
    var date = new Date();
    return date.getUTCHours()+(date.getUTCMinutes()/60)+(date.getUTCSeconds()/3600)
}
function exportToCsv(table){
    var rows=Array.from(table.querySelectorAll("tr"))
    var longest= findLongestRowLength(rows)
    var lines=[]
    var linesString;
    for(var row of rows){
        var line=""
        for(var x=0;x<longest;x++){
            if(row.children[x]!=undefined){
                line+=parseCell(row.children[x])
            }
            line+=(x!=longest-1)?",":""
        }
        lines.push(line)
    }
    linesString=lines.join('\n')
    var csvBlob=new Blob([linesString],{type:'text/csv'})
    var blocUrl =URL.createObjectURL(csvBlob)
    console.log(blocUrl)
    var aElement=document.createElement('a')
    aElement.href=blocUrl
    aElement.download=table.id+'.csv'
    aElement.click()
    setTimeout(()=>{
        URL.revokeObjectURL(blocUrl)
    },500)
    }
    function findLongestRowLength(rows){
    var retNum=0
    for(x in rows){
        retNum=Math.max(rows[x].childElementCount,retNum)
    }
    return retNum
    }
    function parseCell(cell){
    var parsedValue=cell.textContent
    parsedValue=parsedValue.replace(/"/g,'""')
    parsedValue=/[",\n"]/.test(parsedValue)?`"${parsedValue}"`:parsedValue
    return parsedValue
    }
    

function closeSidebar(){
    if(sidebarWidth>0){
        sidebarWidth-=5
        menuSidebar.style.width=sidebarWidth+"px"
        setTimeout(closeSidebar,1)
        let pages = document.querySelectorAll(".page");
        for (let i = 0; i < pages.length; i++) {
            pages[i].style.opacity = 1-(sidebarWidth/300);
        }
    }
    menuSidebar.style.visibility=sidebarWidth>0?"visible":"hidden"
}
function openSidebar(){
    if(sidebarWidth<150){
        sidebarWidth+=5
        menuSidebar.style.width=sidebarWidth+"px"
        setTimeout(openSidebar,1)
        let pages = document.querySelectorAll(".page");
        for (let i = 0; i < pages.length; i++) {
            pages[i].style.opacity = 1-(sidebarWidth/300);
        }
    }
    menuSidebar.style.visibility=sidebarWidth>0?"visible":"hidden"
}



var lastElement={}
var lastBaseName={}
var inverSort={"deviceoverviewTable":false,"useroverviewTable":false}
var table, rows, sorting, i, x, y, shouldSwitch,sortIndex,dataType,sortLoopInterval;

function sortTable(table,sortIndex,isnumber,baseName,element) {  
    if(sorting){
        alert("Sort Error")
    }else{
        try{
            lastElement[table.id].innerHTML=lastBaseName[table.id]
        }catch (ignored) {}
        if(lastElement[table.id]==element){
            inverSort[table.id]=!inverSort[table.id]
        }
        lastElement[table.id]=element
        lastBaseName[table.id]=baseName

        element.innerHTML=baseName+(inverSort[table.id]?"\u25b2":"\u25bc")

        sorting = true;
        this.table=table
        this.sortIndex=sortIndex
        dataType=isnumber
        while(sorting){
        //sortLoopInterval=setInterval(sortLoop,0)
        sortLoop()
        }
    }
}
function sortLoop(){
    rows = table.rows;
    for (i = 1; i < (rows.length - 1); i++) {
        shouldSwitch = false;
        x = rows[i].getElementsByTagName("TD")[sortIndex].innerHTML;
        y = rows[i + 1].getElementsByTagName("TD")[sortIndex].innerHTML;

        if(x.includes("$")){
            x=x.substr(1)
        }
        if(y.includes("$")){
            y=y.substr(1)
        }
        if(dataType==1){
            y=parseFloat(y)
            x=parseFloat(x)
        }else if (dataType==2){
            y=(new Date(y)).getTime()
            x=(new Date(x)).getTime()
        }else{
            x=x.toLowerCase()
            y=y.toLowerCase()
        }
        if ((x < y&&inverSort[table.id])||(x > y&&!inverSort[table.id])) {
            shouldSwitch = true;
            break;
        }
        
    }
    if (shouldSwitch) {
        rows[i].parentNode.insertBefore(rows[i + 1], rows[i]);
    }else{
        // clearInterval(sortLoopInterval)
        sorting=false
    }
}

pytz/tzfile.py
#!/usr/bin/env python
'''
$Id: tzfile.py,v 1.8 2004/06/03 00:15:24 zenzen Exp $
'''

try:
    from cStringIO import StringIO
except ImportError:
    from io import StringIO
from datetime import datetime, timedelta
from struct import unpack, calcsize

from pytz.tzinfo import StaticTzInfo, DstTzInfo, memorized_ttinfo
from pytz.tzinfo import memorized_datetime, memorized_timedelta

def _byte_string(s):
    """Cast a string or byte string to an ASCII byte string."""
    return s.encode('ASCII')

_NULL = _byte_string('\0')

def _std_string(s):
    """Cast a string or byte string to an ASCII string."""
    return str(s.decode('ASCII'))

def build_tzinfo(zone, fp):
    head_fmt = '>4s c 15x 6l'
    head_size = calcsize(head_fmt)
    (magic, format, ttisgmtcnt, ttisstdcnt,leapcnt, timecnt,
        typecnt, charcnt) =  unpack(head_fmt, fp.read(head_size))

    # Make sure it is a tzfile(5) file
    assert magic == _byte_string('TZif'), 'Got magic %s' % repr(magic)

    # Read out the transition times, localtime indices and ttinfo structures.
    data_fmt = '>%(timecnt)dl %(timecnt)dB %(ttinfo)s %(charcnt)ds' % dict(
        timecnt=timecnt, ttinfo='lBB'*typecnt, charcnt=charcnt)
    data_size = calcsize(data_fmt)
    data = unpack(data_fmt, fp.read(data_size))

    # make sure we unpacked the right number of values
    assert len(data) == 2 * timecnt + 3 * typecnt + 1
    transitions = [memorized_datetime(trans)
                   for trans in data[:timecnt]]
    lindexes = list(data[timecnt:2 * timecnt])
    ttinfo_raw = data[2 * timecnt:-1]
    tznames_raw = data[-1]
    del data

    # Process ttinfo into separate structs
    ttinfo = []
    tznames = {}
    i = 0
    while i < len(ttinfo_raw):
        # have we looked up this timezone name yet?
        tzname_offset = ttinfo_raw[i+2]
        if tzname_offset not in tznames:
            nul = tznames_raw.find(_NULL, tzname_offset)
            if nul < 0:
                nul = len(tznames_raw)
            tznames[tzname_offset] = _std_string(
                tznames_raw[tzname_offset:nul])
        ttinfo.append((ttinfo_raw[i],
                       bool(ttinfo_raw[i+1]),
                       tznames[tzname_offset]))
        i += 3

    # Now build the timezone object
    if len(ttinfo) ==1 or len(transitions) == 0:
        ttinfo[0][0], ttinfo[0][2]
        cls = type(zone, (StaticTzInfo,), dict(
            zone=zone,
            _utcoffset=memorized_timedelta(ttinfo[0][0]),
            _tzname=ttinfo[0][2]))
    else:
        # Early dates use the first standard time ttinfo
        i = 0
        while ttinfo[i][1]:
            i += 1
        if ttinfo[i] == ttinfo[lindexes[0]]:
            transitions[0] = datetime.min
        else:
            transitions.insert(0, datetime.min)
            lindexes.insert(0, i)

        # calculate transition info
        transition_info = []
        for i in range(len(transitions)):
            inf = ttinfo[lindexes[i]]
            utcoffset = inf[0]
            if not inf[1]:
                dst = 0
            else:
                for j in range(i-1, -1, -1):
                    prev_inf = ttinfo[lindexes[j]]
                    if not prev_inf[1]:
                        break
                dst = inf[0] - prev_inf[0] # dst offset

                # Bad dst? Look further. DST > 24 hours happens when
                # a timzone has moved across the international dateline.
                if dst <= 0 or dst > 3600*3:
                    for j in range(i+1, len(transitions)):
                        stdinf = ttinfo[lindexes[j]]
                        if not stdinf[1]:
                            dst = inf[0] - stdinf[0]
                            if dst > 0:
                                break # Found a useful std time.

            tzname = inf[2]

            # Round utcoffset and dst to the nearest minute or the
            # datetime library will complain. Conversions to these timezones
            # might be up to plus or minus 30 seconds out, but it is
            # the best we can do.
            utcoffset = int((utcoffset + 30) // 60) * 60
            dst = int((dst + 30) // 60) * 60
            transition_info.append(memorized_ttinfo(utcoffset, dst, tzname))

        cls = type(zone, (DstTzInfo,), dict(
            zone=zone,
            _utc_transition_times=transitions,
            _transition_info=transition_info))

    return cls()

if __name__ == '__main__':
    import os.path
    from pprint import pprint
    # Patched in Debian, use the system zoninfo from the tzdata package
    base = '/usr/share/zoneinfo'
    tz = build_tzinfo('Australia/Melbourne',
                      open(os.path.join(base,'Australia','Melbourne'), 'rb'))
    tz = build_tzinfo('US/Eastern',
                      open(os.path.join(base,'US','Eastern'), 'rb'))
    pprint(tz._utc_transition_times)
    #print tz.asPython(4)
    #print tz.transitions_mapping


index.js
const config = require('config');
const https = require('https');
const http = require('http');
const fs = require('fs');
const url = require('url');
let {PythonShell} = require('python-shell')
var path = require('path');
var numOfDevices=0;
const debug=true;
var count=0;
var timestamp=-1;

var serverOptions=null

if (getConfig("enableHttps")) {
    serverOptions={
        key: fs.readFileSync('key.pem'),
        cert: fs.readFileSync('cert.pem')
    }
}
const server = http.createServer(null,async (req, res) => {
    var ip = req.headers['x-forwarded-for'] ||      req.connection.remoteAddress ||      req.socket.remoteAddress ||     (req.connection.socket ? req.connection.socket.remoteAddress : null);
    appendLog("http req: "+req.url+" From Ip address: "+ip)
    var filePath = './www' + decodeURI(req.url).split('?')[0];
    if(filePath.includes("dashboard")&&!getConfig("enableDashboard")){
        filePath="./www/"
    }
    if (filePath == './www/'){
        filePath = getConfig("enableDashboard")?'./www/dashboard.html':'./www/HoneyGainDevicesDelta_plot.html';
    }
    var extname = path.extname(filePath);
    var contentType = 'text/html';
    switch (extname) {
        case '.js':
            contentType = 'text/javascript';
            break;
        case '.css':
            contentType = 'text/css';
            break;
        case '.json':
            contentType = 'application/json';
            break;
        case '.png':
            contentType = 'image/png';
            break;      
        case '.jpg':
            contentType = 'image/jpg';
            break;
        case '.wav':
            contentType = 'audio/wav';
            break;
    }
    if(filePath!="./www/styles.css"&&false&&getConfig("requireLogin")){
        filePath="./www/login.html"
    }
    if(filePath=="./www/dashboard/getdata"){
        var url_parts = url.parse(req.url, true);
        var query = url_parts.query;
        var responseText="{Error:No action Defined}"
        if(query!=undefined&&query.action!=undefined){
            switch(query.action){
                case "getbalance":
                    var x = await sendRequest("/api/v1/users/balances");
                    responseText='{"req":'+x+',"echo":"'+query.echo+'"}'
                break;
                case "getdevicebalance":
                    content=await wsReadFile("data.json")
                    jsonData=JSON.parse('{"dataFile":['+content.slice(0,-2)+']}')
                    var retBal={}
                    for (x in jsonData.dataFile){
                        var date=jsonData.dataFile[x]['9']
                        var id=jsonData.dataFile[x]['1']
                        var credits=jsonData.dataFile[x]['8']
                        if(query.time!="now"&&query.starttime!=undefined&&date>query.starttime){
                            if(retBal[id]==undefined){
                                retBal[id]=credits
                            }
                        }
                        if(query.time!="now"&&query.endtime!=undefined&&date<query.endtime){
                            retBal[id]={"credits":credits}
                        }
                        if(query.time=="now"){
                            if(retBal[id]!=undefined){
                                if(credits>retBal[id].credits){
                                    retBal[id]={"credits":credits,"lastEarning":date}
                                }
                            }else{
                                retBal[id]={"credits":credits,"lastEarning":date}
                            }
                        }
                    }
                    responseText='{"balance":'+JSON.stringify(retBal)+',"time":"'+(query.starttime||query.endtime)+'","echo":"'+query.echo+'"}'
                break;
                case "getstarttime":
                    content=await wsReadFile("data.json")
                    jsonData=JSON.parse('{"dataFile":['+content.slice(0,-2)+']}')
                    var retTime=(new Date()).getTime()/1000
                    for (x in jsonData.dataFile){
                        retTime=Math.min(retTime,jsonData.dataFile[x]['9'])                                    
                    }
                    responseText='{"time":"'+retTime+'","echo":"'+query.echo+'"}'
                break;
                case "getdata":
                    content=await wsReadFile("data.json")
                    responseText='{"dataFile":['+content.slice(0,-2)+'],"echo":"'+query.echo+'"}'
                break;
                case "getlastbalance":
                    content=await wsReadFile("lastbalance.json")
                    content2=await ReadFile("lastlastbalance.json",content)
                    responseText='{"req":'+content+',"req2":'+content2+',"echo":"'+query.echo+'"}'
                break;
                case "getidmap":
                    content=await wsReadFile("idmap.json")
                    responseText='{"idmap":'+content+',"echo":"'+query.echo+'"}'
                break;
                case "gettransactions":
                    getTransactions(1)
                    try{
                        transactionsArray=JSON.parse(await ReadFile('transactions.json','[]'))
                    }catch{
                        console.log('Cannot Parse File: transactions.json')
                    }                       
                    var retData=[]
                    for (var transaction of transactionsArray){
                        var transactionDate=(new Date(transaction.created_at)).getTime()
                        if(transactionDate>query.starttime&&transactionDate<query.endtime){
                            retData.push(transaction)
                        }
                    }
                    responseText='{"transactions":'+JSON.stringify(retData)+',"echo":"'+query.echo+'"}'
                break;
            }
        }
        res.writeHead(200, { 'Content-Type': 'application/json'});
        res.end(responseText, 'utf-8');
    }else{
        fs.readFile(filePath, function(error, content) {
            if (error) {
                if(error.code == 'ENOENT'){
                    res.writeHead(404, { 'Content-Type': 'text/plain' });
                    res.end("Error 404: File "+filePath+" Not Found", 'utf-8');
                }
                else {
                    res.writeHead(500);
                    res.end('check with the site admin for error: '+error.code+' ..\n');
                    res.end(); 
                }
            }
            else {
                res.writeHead(200, { 'Content-Type': contentType});
                res.end(content, 'utf-8');
            }
        });
    }
});
server.listen(getConfig("webserverPort"),getConfig("webserverHost"), (h,p) => {
    console.log("Server running at http://"+getConfig("webserverHost")+":"+getConfig("webserverPort")+"");
});

function wsReadFile(fPath){
    return new Promise(resolve => {
        fs.readFile(fPath, function(error, content) {
            if (error) {
                if(error.code == 'ENOENT'){
                    //connection.sendUTF("Error 404: File "+fPath+" Not Found");
                    appendLog("Error 404: File "+fPath+" Not Found")
                }
                resolve(undefined)
            }
            else {
                resolve(content)
            }
        });
    })
}
function ReadFile(fPath,onFail){
    return new Promise(resolve => {
        fs.readFile(fPath, function(error, content) {
            if (error) {
                if(error.code == 'ENOENT'){
                    appendLog("Error 404: File "+fPath+" Not Found")
                }
                resolve(onFail)
            }
            else {
                resolve(content)
            }
        });
    })
}
function appendLog(logStr){
    console.log(logStr);
        fs.appendFile("latest.log", logStr+"\n", function(err) {
            if(err) {
                return console.log(err);
            }
        }); 
}
function genGraph(){
    if(debug){
        appendLog("genGraph")
    }
    var options={
        args:[getConfig("graphTitle")]
    }
    PythonShell.run('HoneyGainDevicesDelta_plot.py', options, function (err, results) {
      if (err) 
        throw err;
        if(debug){
            appendLog('results: '+results);
        }
    });
  }

function sendRequest(url){
    return new Promise(resolve => {
        var options = {
            hostname: 'dashboard.honeygain.com',
            port: 443,
            path: url,
            method: 'GET',
            headers: {
                'authorization': 'Bearer '+getConfig("authToken")
            }
        }
        if(debug){
            appendLog("sendRequest")
            appendLog("URL: "+options.path)
        }
        callback = function(response) {
            let str=''
            response.on('data', function (chunk) {
                str += chunk;
            });

            response.on('end', function () {
                if(response.statusCode==200){
                    resolve(str)
                }else{
                    appendLog("HTTP Get Error: "+str);
                }
            });
        }

        var req = https.request(options, callback).end();
        req.on('error', error => {
            appendLog("HTTP Get "+error)
        });
    });
}
function getConfig(keyName){
    var value;
    try{
        value=config.get(keyName)
    }catch(e){
        var defaultConfig=
        {
            "authToken":"<Insert Your Bearer Token>",
            "pingInterval":3600,
            "startOnTheHour":true,
            "graphTitle":"Honey Gain Devices Delta",
            "enableDashboard":true,
            "enableGraph":true,
            "useHTTPAuth":false,
            "webserverPort":80,
            "webserverHost":"127.0.0.1",
            "enableHttps": false,
            "requireLogin":false
        }
        if(defaultConfig[keyName]==undefined){
            throw Error(e);
        }else{
            value=defaultConfig[keyName];
            appendLog("Key: "+keyName+" Not found in config using default value: "+value);
        }
    }
    return value
}
async function getNumOfDevices(){
    timestamp=Math.round(Date.now() / 1000)
    var x= await sendRequest('/api/v1/users/me');
    var jsonData=JSON.parse(x).data;
    numOfDevices=jsonData.total_devices;
    if(count==0){
        appendLog("Total Devices: "+numOfDevices)
    }
   
}
async function getDevices(pageNum){
    var res=await sendRequest('/api/v1/devices'+(pageNum==1?'':'?page='+pageNum))
    var jsonData=JSON.parse(res).data
    var idMap
    try{
        idMap=JSON.parse(await ReadFile("idmap.json","{}"))
    }catch{

    }
    if(idMap==undefined){
        idMap={}
        appendLog("idmap.json could not be parsed")
    }
    for(var device in jsonData){
        if(idMap[jsonData[device].id]==undefined||idMap[jsonData[device].id].title!=encodeURI(jsonData[device].title)){
            var idmapData={
                id:jsonData[device].id,
                manufacturer:encodeURI(jsonData[device].manufacturer).replace(/'/g, "%27"),
                model:encodeURI(jsonData[device].model).replace(/'/g, "%27"),
                title:encodeURI(jsonData[device].title).replace(/'/g, "%27"),
                platform:encodeURI(jsonData[device].platform).replace(/'/g, "%27"),
                version:encodeURI(jsonData[device].version).replace(/'/g, "%27"),
            }
            if(debug){
                appendLog("device not found or changed adding: "+JSON.stringify(idmapData))
            }
            idMap[jsonData[device].id]=idmapData
        }
        var deviceData={
            1:jsonData[device].id,
            // 2:encodeURI(jsonData[device].manufacturer),
            // 3:encodeURI(jsonData[device].model),
            // 4:encodeURI(jsonData[device].title),
            // 5:encodeURI(jsonData[device].platform),
            // 6:encodeURI(jsonData[device].version),
            7:jsonData[device].stats.total_traffic,
            8:jsonData[device].stats.total_credits,
            9:timestamp,
        };
        fs.appendFile("data.json", JSON.stringify(deviceData)+",\n", function(err) {
            if(err) {
                return console.log(err);
            }
        }); 
    }
    fs.writeFile("idmap.json", JSON.stringify(idMap, null, 2), function(err) {
        if(err) {
            return console.log(err);
        }
    }); 
    if(pageNum<numOfDevices/10){
        console.log("Pagenum: "+pageNum)
        pageNum++;
      getDevices(pageNum);
    }else if (getConfig("enableGraph")){
        genGraph();
    }
}
var splitTransactionArray
async function getTransactions(pageNum,inArray){
    var res=await sendRequest('https://dashboard.honeygain.com/api/v1/transactions?page='+pageNum)
    var resJson=JSON.parse(res)
    var resArray=resJson.data
    var totalPages=resJson.meta.pagination.total_pages
    var totalItems=resJson.meta.pagination.total_items
    var transactionsArray=[]
    if(pageNum==1){
        splitTransactionArray=false
        try{
            transactionsArray=JSON.parse(await ReadFile('transactions.json','[]'))
        }catch{
            console.log('Cannot Parse File: transactions.json')
        }
    }else{
        transactionsArray=inArray
    }
    if(totalItems==transactionsArray.length){
        if(debug){
            console.log("totalItems length match not updating transactions")
        }
        return
    }
    if(totalItems<transactionsArray.length&&pageNum==1){
        if(debug){
            console.log("totalItems length to high resetting transactions")
        }
        transactionsArray=[]
    }
    resArray.reverse()
    var index=0
    for(var newData of resArray){
        var found=-1
        for(var aIndex in transactionsArray){
            if(transactionsArray[aIndex].id==newData.id){
                found=aIndex
            }
        }
        if(found==-1&&index==0){
            if(pageNum<totalPages){
                transactionsArray=await getTransactions(pageNum+1,transactionsArray)
            }
        }
        if(!splitTransactionArray){
            if(found>-1){
                splitTransactionArray=true
                transactionsArray.splice(found)
            }else if(pageNum>=totalPages){
                splitTransactionArray=true
                transactionsArray=[]
            }
        }
        transactionsArray.push(newData)
        index++
    }

    if(pageNum==1){
        var outString='[\n'
        for(var x of transactionsArray){
            outString+=JSON.stringify(x)+',\n'
        }
        outString=outString.substr(0,outString.length-2)
        outString+='\n]'
        fs.writeFile("transactions.json", outString, function(err) {
            if(err) {
                return console.log(err);
            }
        }); 
    }else{
        return transactionsArray
    }
}
async function getBalance(){
    var content=await sendRequest("/api/v1/users/balances");
    var lasBalance=await ReadFile("lastbalance.json",content)
    fs.writeFile("lastbalance.json", content, function(err) {
        if(err) {
            return console.log(err);
        }
    }); 
    fs.writeFile("lastlastbalance.json", lasBalance, function(err) {
        if(err) {
            return console.log(err);
        }
    }); 
}
function main(){
    getNumOfDevices()
    getDevices(1)
    getTransactions(1)
    getBalance()
}
function init(){
    setInterval(main,getConfig("pingInterval")*1000) 
    main()
}
var passedIntervals=parseInt(((new Date).getMinutes()*60+(new Date).getSeconds())/getConfig("pingInterval"))
var nextIntervalTime=(passedIntervals+1)*getConfig("pingInterval")
var secondsToNextInterval=nextIntervalTime-((new Date).getMinutes()*60+(new Date).getSeconds())
setTimeout(init,secondsToNextInterval*1000)
getTransactions(1)
fs.readFile("lastbalance.json", function(error, content) {
    if (error) {
        getBalance()
    }
});